app/src/views/private/components/export-sidebar-detail.vue

```diff
@@ -386,14 +377,50 @@ const getItemCount = debounce(async () => {
 		}
 
 		itemCount.value = count;
+		return count;
 	} finally {
 		itemCountLoading.value = false;
 	}
-}, 250);
+};
+
+watch([() => exportSettings.search, () => exportSettings.filter], debounce(getItemCount, 250));
+
+watch(
+	() => exportSettings.limit,
+	() => {
+		if (exportSettings.limit < -1) {
+			exportSettings.limit = -1;
+		}
+
+		if (
+			exportSettings.limit !== null &&
+			!Number.isNaN(exportSettings.limit) &&
+			!Number.isInteger(exportSettings.limit)
+		) {
+			exportSettings.limit = Math.round(exportSettings.limit);
+		}
+	}
+);
+
+const exportCount = computed(() => {
+	const limit = exportSettings.limit === null || exportSettings.limit === -1 ? Infinity : exportSettings.limit;
+	return itemCount.value !== undefined ? Math.min(itemCount.value, limit) : limit;
+});
+
+watch(exportCount, () => {
+	const queryLimitThreshold = exportCount.value > queryLimitMax;
+	const batchThreshold = exportCount.value >= 2500;
 
-watch(exportSettings, () => {
-	if (exportDialogActive.value) {
-		getItemCount();
+	if (queryLimitThreshold || batchThreshold) {
+		lockedToFiles.value = {
+			lockedBy: queryLimitThreshold ? 'queryLimitThreshold' : 'batchThreshold',
+			previousLocation: lockedToFiles.value?.previousLocation ?? location.value,
+		};
+
+		location.value = 'files';
+	} else if (lockedToFiles.value) {
+		location.value = lockedToFiles.value.previousLocation;
+		lockedToFiles.value = null;
 	}
```

Comment:
This logic should also run upon load to correctly set the initial location and disabling download.